### PR TITLE
Prevent losing focus when saving on media modal

### LINKED ISSUE #197: ISC form elements defocus when Ajax saving call completes in Media Library modal
In the Media Library modal—both in _Media > Library_ as well as in a post, selecting images from the Media Library modal—, after changing the content of ISC form element “A” (input field, checkbox, dropdown menu) of any image and tabbing or clicking into ISC form element “B” triggers an Ajax call to save the info of form element “A“. Once this call is completed (positive HTTP feedback, I reckon) the presently focussed form element “B” becomes defocussed.
This does _not_ affect ISC Pro’s batch editing of form elements in the Media Library list view. Neither does it affect the WP core input fields “Alternative text”, “Caption”, etc. in the modal.

How to reproduce
---

1. Go either to _Media > Library (grid view)_ or edit a random post and place an image block, enter the Media Library modal
2. Select a random image, put cursor in input field “Image Source”, enter random info
3. Click or (speedily) tab into input field “Image Source URL” and watch it closely
4. After a second or two, depending on your ping and the server’s load, the “Image Source URL” input field becomes defocussed, the cursor is displaced from it

Expected behavior
---

The expected behavior is for the defocussing to seize, for any constellation of the exact form elements “A” and “B”.

Priority
---

Medium. This has been a minor problem for a while, but now, a [first report of annoyance](https://wordpress.org/support/topic/image-source-not-showing-on-picture/#post-15739835) has been submitted.

Additional context
---

I think I remember that this had been a problem for the post-editing mode’s **sidebar** ISC input fields, too, but this seems to be gone after the switch to REST API calls in ISC 2.7.0.

Tests
---

- [ ] forwards tabbing, backwards tabbing and clicking from form element “A” to “B”—for any constellation of the four ISC form elements—does not defocus form element “B” after automatic background saving of “A”

### PR DESCRIPTION
**Pull request by cnhk**
When editing attachment fields in the media modal, focus is lost after saving if the field is entered by tabbing or clicking on it when the previously edited field still has focus

fix #197

### CODE REVIEW

**Comment by webzunft** [comment]
@cnhk thanks. I wasn’t aware that this also affected the block editor. @Galanos if you were then please also test it.

**Comment by cnhk** [comment]
It's happening in every page that can possibly load the media library modal

**Review by Galanos — CHANGES_REQUESTED** [review]
I checked the two places that I know of:
- Media > Library > (select grid view) > (select image) > (sidebar in modal)
- Posts > All Posts > (create post) > (insert image) > (select button “Media Library” in empty image block) > (sidebar in modal)

Works like a charm—tabbing forwards, tabbing backwards, clicking into next field, clicking into license dropdown, … all is well.
The entered ISC metadata is saved to the attachment just as expected.

But the fix has spawned a new issue. It’s easiest to explain the replication steps:
1. Create a post, place an image, save the post, leave the post.
2. Re-enter the post.
3. Select the image, select “Replace” in the toolbar, select “Open Media Library”, Media Library modal pops up.
4. Without clicking anything, scroll down the sidebar of the preselected image.
5. Observe the ISC form fields missing.

The following applies:
- this does not happen with the current master branch
- selecting a different image, then reselecting the original image makes its ISC form fields appear
- _not_ selecting a different image, but leaving and re-entering the modal makes the original image’s ISC form fields appear, too
- to replicate once more, you need to leave and re-enter the post
- since in the “Media > Library” modal there’s no image preselected, it doesn’t apply there

**Comment by cnhk** [comment]
I can't find an ideal (an not too complex) solution for this. First of all, I can't get the other solution in the core issue ticket to work, not enough info, so I dropped it.

So

1. We can tweak the current solution a bit, and apply it only on the `upload.php` page. When accessing the media modal frame from elsewhere (post page widgets etc.) the losing focus bug will be present.
2. Another approach is to return the focus manually to the input if it had it when WordPress re-render the attachment fields after a save event. But we do that only with ISC fields (no need to risk problems with others' attachment custom fields). Also if one is typing really fast in a textfield while the re-render process occurs, there might be a kind of stutter and one letter my be just swallowed. This is just in theory, I didn't start implementing yet.

**Comment by cnhk** [comment]
@galanos I kind of created a mix of the two approaches from my previous comment. Please test the new fix. It prevents re-rendering custom attachment fields more than once. The potential side effect is that it can break the behavior of some custom fields, but only when it is really a very very complicated field (something coupled with a lot of complex custom JavaScript).

**Comment by Galanos** [comment]
Note to self: Check freshly uploaded images for ISC form fields in the media library modal’s sidebar.

**Comment by Galanos** [comment]
@cnhk I started testing and noticed … 
- when replacing an image with itself, changes made to its ISC fields in the Media Library modal within the block editor do reflect in the Media Library (opened in another tab), but **not** in the post’s block editor sidebar
- saving the post at this point replaces the ISC metadata with the unmodified, wrong sidebar values
- not saving the post and re-entering loads the modified, correct sidebar values for the image

I fear that this might be hard to read, so I’ll list the steps:
1. In a post, replace an image with ISC info through the toolbar’s “Replace” button.
2. In the modal, keep the preselected image selected, modify its ISC info.
3. Confirm the changes with the “Select” button, the modal closes, do not save post yet.
4. See unmodified ISC info in the sidebar, do not save post yet.
5. See modified ISC info in Media Library in second browser tab.
6. Save the post.
7. Refresh and see unmodified ISC info in Media Library.

But in the modal called from the block editor, the ISC fields appear for the preselected image and the focussing problem is still gone, though :)

**Comment by cnhk** [comment]
@Galanos This is an unrelated issue. It is happening since we use WP's core functions to load/save ISC fields ([this](https://github.com/image-source-control/image-source-control/pull/187)) instead of custom AJAX calls.

Also, if we fix that bug (block sidebar on post/page after an edit in the modal), the user will be required to re-save the post/page since we did a modification on one or more attributes of a block on that page (not a big deal I think but worth mentioning).

**Review by Galanos — APPROVED** [review]
Given that the [last issue that I had raised](https://github.com/image-source-control/image-source-control/pull/199#issuecomment-1177423807) had no connection to the original focussing issue and that the latest commit was tested and found to solve the focussing _plus_ the missing sidebar ISC section, I have no objections to merging.

@cnhk Would you be so kind to please solve the remaining ReadMe conflict?

**Comment by cnhk** [comment]
@Galanos Done

File: admin/admin.php [hunk 1]
```diff
@@ -117,12 +117,19 @@ public function admin_notices() {
 	 */
 	public function add_admin_scripts( $hook ) {
 		$screen = get_current_screen();
-		if ( isset( $screen->id ) && $screen->id === 'settings_page_isc-settings' ) {
-			wp_enqueue_script( 'isc_settings_script', plugins_url( '/assets/js/settings.js', __FILE__ ), array(), ISCVERSION, true );
-		}
 
-		if ( isset( $screen->id ) && $screen->id === 'media_page_isc-sources' ) {
-			wp_enqueue_script( 'isc_sources_script', plugins_url( '/assets/js/sources.js', __FILE__ ), array(), ISCVERSION, true );
+		if ( isset( $screen->id ) ) {
```
Comment (webzunft) [inline admin/admin.php:121]: you could move the check to the top and return if it is false, so no brackets are needed.

**Review by webzunft — CHANGES_REQUESTED** [review]

File: admin/admin.php (hunk 1 above)
Comment (webzunft) [reply admin/admin.php:121]: @cnhk ping

**Comment by Galanos** [comment]
@cnhk I’m a bit confused—I installed the current state of branch “i-197” on a staging site but the defocussing problem is back just as it was when we started. If I didn’t mix something up while downloading or installing (I did both twice to make sure), something must have gone wrong while committing between early July and now.
Could you please see for yourself? Thank you.

**Comment by cnhk** [comment]
@Galanos My bad, I forgot one single `!` in https://github.com/image-source-control/image-source-control/pull/199/commits/234975d83c5b08640ea0b133dc88e2fa425569dc

Fixed now.

**Pull request merged**